enhancements/network/router-ca-configmap.md

```diff
@@ -0,0 +1,219 @@
+
+* [openshift/cluster-ingress-operator#109 "Use self-signed default router certificate"](https://github.com/openshift/cluster-ingress-operator/pull/109)
+  added generation of default certificates and introduced the `router-ca`
+  ConfigMap.
+* [openshift/cluster-kube-apiserver-operator#222 "Trust cluster-ingress-operator's CA certificate"](https://github.com/openshift/cluster-kube-apiserver-operator/pull/222)
+  proposed adding `router-ca` to kube-apiserver's trust bundle; it was not
+  merged.  During review of this PR, it was determined that the ingress operator
+  must conditionally publish `router-ca`.
+* [openshift/cluster-kube-controller-manager-operator#145 "Trust cluster-ingress-operator's CA certificate"](https://github.com/openshift/cluster-kube-controller-manager-operator/pull/145)
+  added `router-ca` to kube-controller-manager's trust bundle.
+* [openshift/cluster-ingress-operator#110 "Put router-ca configmap in openshift-config-managed iff needed"](https://github.com/openshift/cluster-ingress-operator/pull/110)
+  changed the ConfigMap's namespace to `openshift-config-managed` and made its
+  publication conditional on its use.
+* [openshift/cluster-ingress-operator#142 "controller/certificate: relocate CA publishing"](https://github.com/openshift/cluster-ingress-operator/pull/142)
+  moved the `router-ca` publication to a separate controller.
+* [openshift/console-operator#328 "Bug 1764704: Sync router-ca to the console namespace"](https://github.com/openshift/console-operator/pull/328)
+  added `router-ca` to OpenShift Console's trust bundle.
+* [openshift/cluster-ingress-operator#329 "Always publish router-ca configmap"](https://github.com/openshift/cluster-ingress-operator/pull/329)
+  proposes making publication unconditional.
+
+## Drawbacks
+
+Publishing the ingress operator's signing certificate does not help
+if the administrator configures a custom default certificate.
+
+## Alternatives
+
+### Add a Cluster Configuration API for the Ingress CA
+
+Instead of using a ConfigMap, the ingress CA should be formal cluster
+configuration in the `spec` and `status` fields of the `default` cluster
+`ingresses.config.openshift.io` resource.  The `spec` field would be used to
+configure a custom ingress CA, and the `status` field would report the effective
+ingress CA.
+
+#### Drawbacks
+
+This approach would add a step to configuring a custom default certificate with
+custom PKI.
+
+The open questions around the semantics of multiple IngressControllers still
+need to be addressed with this approach.
+
+Network Edge has proposed this approach in the past and achieved no traction
+with it.
+
+### Inject the Ingress CA into ConfigMaps with a Well Known Label
+
+Instead of or in addition to publishing the `router-ca` ConfigMap, the ingress
+operator could inject the ingress CA into all ConfigMaps with a specific label,
+for example `config.openshift.io/inject-ingress-cabundle: "true"`.  This approach
```

Comment:
Instead of having the ingress operator inject the ingress ca cert, why not have the existing [trust bundle publishing controller](https://github.com/openshift/cluster-network-operator/blob/master/pkg/controller/proxyconfig/controller.go#L202-L310) handle this? The controller can be expanded to watch for the `router-ca` configmap, add the cert to the cluster-wide trusted ca bundle configmap and have the combined bundle injected into components. Most of this workflow is already implemented and this approach would allow a single operator to manage cluster-wide trust bundle publishing and injection. All components that rely on routes to verify access, should already be setup for the trust bundle injection so no additional changes should be required.

Reply:
I like the idea of re-using an existing controller.  However, we would still need a distinct annotation because components that need to trust the proxy CA may not need to trust `router-ca`, and we must maintain separation of chains of trust.

Reply:
ack, I get your point and agree. I did a review of the the ingress operator cert code and I see the `router-ca` configmap is only created when it's the self generated cert (i.e. no ingress controllers define `defaultCertificate`). Why does the ingress operator publish the self generated cert and not user-defined certs?

Reply:
The operator publishes the CA that it uses to sign operator-generated default certificates.  If a custom default certificate is provided, we have the certificate but not the signing CA for that certificate (unless the administrator happens to include the CA in the default certificate secret, which we cannot assume to be the case).  What @deads2k has proposed is to publish the custom default certificate itself as if it were the CA, which is technically wrong but will work for Go-based clients.